tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,382 @@
+
+        x = x - int((self.patch_size[1] - 1) / 2)
+        y = y - int((self.patch_size[0] - 1) / 2)
+
+        data = self.wsi.read_rect(location=(x, y), size=self.patch_size)
+
+        return data
+
+    # def extract_patches(self, input_img, labels=None):
+    #     if not isinstance(labels, np.ndarray):
+    #         raise Exception("Please input correct csv, json path or csv data")
+    #
+    #
+    #     patch_h = self.img_patch_h
+    #     patch_w = self.img_patch_w
+    #
+    #     image = np.lib.pad(
+    #         image,
+    #         ((self.pad_y, self.pad_y), (self.pad_x, self.pad_x), (0, 0)),
+    #         "symmetric",
+    #     )
+    #
+    #     num_patches_img = len(labels) * self.num_examples_per_patch
+    #     img_patches = np.zeros(
+    #         (num_patches_img, patch_h, patch_w, image.shape[2]), dtype=image.dtype
+    #     )
+    #     labels = [None] * num_patches_img
+    #     class_id = [None] * num_patches_img
+    #
+    #     cell_tot = 1
+    #     iter_tot = 0
+    #     for row in labels:
+    #         patch_label = row[0]
+    #         cell_location = np.array([row[2], row[1]], dtype=np.int)
+    #         cell_location[0] = (
+    #             cell_location[0] + self.pad_y - 1
+    #         )  # Python index starts from 0
+    #         cell_location[1] = (
+    #             cell_location[1] + self.pad_x - 1
+    #         )  # Python index starts from 0
+    #         if self.num_examples_per_patch > 1:
+    #             root_num_examples = np.sqrt(self.num_examples_per_patch)
+    #             start_location = -int(root_num_examples / 2)
+    #             end_location = int(root_num_examples + start_location)
+    #         else:
+    #             start_location = 0
+    #             end_location = 1
+    #
+    #         for h in range(start_location, end_location):
+    #             for w in range(start_location, end_location):
+    #                 start_h = cell_location[0] - h - int((patch_h - 1) / 2)
+    #                 start_w = cell_location[1] - w - int((patch_w - 1) / 2)
+    #                 end_h = start_h + patch_h
+    #                 end_w = start_w + patch_w
+    #                 labels[iter_tot] = patch_label
+    #                 class_id[iter_tot] = cell_tot
+    #              img_patches[iter_tot, :, :, :] = image[start_h:end_h, start_w:end_w]
+    #                 iter_tot += 1
+    #
+    #         cell_tot += 1
+    #     return img_patches, labels, class_id
+
+    def merge_patches(self, patches=None):
+        """Merge patch is not supported by :obj:`PointsPatchExtractor`.
+        Calling this function for :obj:`PointsPatchExtractor` will raise an error. This
+        overrides the merge_patches function in the base class :obj:`PatchExtractor`
+
+        """
+        raise MethodNotSupported(
+            message="Merge patches not supported for PointsPatchExtractor"
+        )
+
+
+def read_point_annotations(input_table):
```

Comment:
Move to misc.py

Reply:
Added in this commit https://github.com/TIA-Lab/tiatoolbox/commit/b72516bc3f1436a317256989fd5f4f93b410ed91